src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            IEnumerable<TSource> big = null;
+            IEnumerable<TSource> small = null;
 
-            return IntersectIterator(first, second, null);
+            if (first.Count() > second.Count())
```

Comment:
Count can be O(n), so this could be on a similar order in itself as Intersect was before the change, depending on the source. We do have optimisations that obtain counts only when get can be reasonably fast when that offers a possible optimisation. Examine how they work.